src/ProcedureMigration.php

```diff
@@ -0,0 +1,211 @@
+     */
+    private function __construct(
+        string $name,
+        string $operation,
+        bool $deterministic = false,
+        string $readsSqlData = 'READS SQL DATA',
+        array $inParameterOperations = [],
+        array $outParameterOperations = [],
+        string $body = null
+    ) {
+        $this->name = $name;
+        $this->operation = $operation;
+        $this->deterministic = $deterministic;
+        $this->readsSqlData = $readsSqlData;
+        $this->inParameterOperations = $inParameterOperations;
+        $this->outParameterOperations = $outParameterOperations;
+        $this->body = $body;
+    }
+
+    /**
+     * Pushes a new add column operation.
+     *
+     * @param string $body
+     * @return ProcedureMigration
+     */
+    public function withBody(string $body): ProcedureMigration
+    {
+        if ($this->operation !== ProcedureOperation::CREATE) {
+            throw new LogicException('Procedure body can only be set in a procedure create migration.');
+        }
+
+        $this->body = $body;
+
+        return $this;
+    }
+
+    /**
+     * Returns the table operation.
+     *
+     * @return ProcedureOperation
+     */
+    public function getOperation(): ProcedureOperation
+    {
+        return new ProcedureOperation(
+            $this->name,
+            $this->operation,
+            $this->deterministic,
+            $this->readsSqlData,
+            $this->inParameterOperations,
+            $this->outParameterOperations,
+            $this->body
+        );
+    }
+
+    /**
+     * Sets determinism for the function.
+     *
+     * @param bool $deterministic
+     * @return $this
+     */
+    public function isDeterministic(bool $deterministic): ProcedureMigration
+    {
+        if ($this->operation === ProcedureOperation::DROP) {
+            throw new LogicException('Cannot set deterministic property in a procedure drop migration.');
+        }
+
+        $this->deterministic = $deterministic;
+
+        return $this;
+    }
+
+    /**
+     * Sets the reads sql data property of the function.
+     *
+     * @param string $readsSqlData
+     * @return $this
+     */
+    public function readsSqlData(string $readsSqlData): ProcedureMigration
+    {
+        if ($this->operation === ProcedureOperation::DROP) {
+            throw new LogicException('Cannot set readsSqlData property in a procedure drop migration.');
+        }
+
+        if (!in_array($readsSqlData, ['CONTAINS SQL', 'NO SQL', 'READS SQL DATA', 'MODIFIES SQL DATA'])) {
+            throw new LogicException('Cannot set readsSqlData, not a valid option.');
+        }
+
+        $this->readsSqlData = $readsSqlData;
+
+        return $this;
+    }
+
+    /**
+     * Pushes a new IN add parameter operation.
+     *
+     * @param string $name
+     * @param array  $options
+     * @return $this
+     */
+    public function addInParameter(string $name, array $options = []): ProcedureMigration
```

Comment:
In keeping with the other factory methods, should these methods be called `withInParameter` and `withOutParameter` respectively?

Reply:
I think add is correct for the use, but could add new methods for withInParameters or withOutParameters in the future =)

Reply:
On second though with may make things easier; as long as with is always appending rather than settings lists.